lib/src/utils.dart

```diff
@@ -29,3 +29,64 @@ void validateErrorArgs(
         'the string.');
   }
 }
+
+// See https://en.wikipedia.org/wiki/UTF-16#Code_points_from_U+010000_to_U+10FFFF
+// for documentation on how UTF-16 encoding works and definitions of various
+// related terms.
+
+/// The lower bound of Unicode's supplementary plane.
+const _supplementaryPlaneLowerBound = 0x10000;
+
+/// The lower bound of the UTF-16 high surrogate block.
+const _highSurrogateLowerBound = 0xD800;
+
+/// The lower bound of the UTF-16 low surrogate block.
+const _lowSurrogateLowerBound = 0xDC00;
+
+/// The number of low bits in each code unit of a surrogate pair that goes into
+/// determining which code point it encodes.
+const _surrogateBits = 10;
+
+/// A bit mask that covers the lower bits of a code point, which are then
+/// encoded into its low surrogate.
+const _lowSurrogateMask = (1 << _surrogateBits) - 1;
```

Comment:
If  this value is reused for other things (which it can) maybe name `_surrogateValueMask` (as opposed to `_surrogateTagMask` for the upper six bits, see below). Then it can be used as a mask for the value bits of any surrogate, whether high or low.